aliyun/log/consumer/heart_beat.py

```diff
@@ -28,17 +27,28 @@ def __init__(self, log_client, heartbeat_interval):
         self.mheart_shards = []
         self.shut_down_flag = False
         self.lock = RLock()
+        self.heartbeat_failed_times = 0
         self.logger = HeartBeatLoggerAdapter(
             logging.getLogger(__name__), {"heart_beat": self})
 
     def run(self):
         self.logger.info('heart beat start')
         while not self.shut_down_flag:
             try:
-                response_shards = []
                 last_heatbeat_time = time.time()
 
-                self.log_client.heartbeat(self.mheart_shards, response_shards)
+                with self.lock:
+                    heartbeat_sucessed = self.log_client.heartbeat(self.mheart_shards, self.mheld_shards)
+                if self.heartbeat_failed_times == 1 and heartbeat_sucessed:
+                    self.heartbeat_failed_times = 0
+                if not heartbeat_sucessed:
+                    self.heartbeat_failed_times += 1
+                    if self.heartbeat_failed_times == 2:
+                        with self.lock:
+                            self.mheld_shards = []
+                with self.lock:
+                    self.mheld_shards = list(set(self.mheld_shards))
+                    response_shards = self.mheld_shards
                 self.logger.debug('heart beat result: %s get: %s',
```

Comment:
still need to check `heartbeat_sucessed` or else it will unload all owned shard directly.